### PR TITLE
Expose Port 8000 for easy testing

### PR DESCRIPTION
**Pull request by madhur-tandon**
Exposing the Port 8000 enables us to start a server within the docker environment. This is helpful in testing the ``pyodide`` builds running within the docker environment using external browser programs of the same machine.

### CODE REVIEW

**Comment by madhur-tandon** [comment]
Yes, indeed. But let's consider this use case.
You clone the ``pyodide`` repository and check out the commit ``191ca2c`` with the title ``Bulk of initial work.`` and set up and build the repository referring to the ``README`` of that commit.

When you run ``python -m http.server`` outside the Docker environment (to observe the console output in the web browser), the console gives an error of ``unable to get the local encoding``. Basically, the error comes because the build and the set up of the repository was not done properly in your machine. In a discussion with @mdboom in the gitter channel, these little niggles are the reason they switched to Docker.

It then makes sense to use the docker environment and run the server within the docker environment. 
(since the build was correctly done there). But to manually observe the output in the console, the port needs to be exposed.

Basically, This is done so that some manual observation of outputs in the console is possible. This PR is not meant for automated testing but for manual testing i.e. for people to successfully build the project using Docker (as individual machines and set ups give problems) and then verify the working in a web browser using a server started within the docker.

Please let me know if this is feasible.
Thanks

**Comment by rth** [comment]
I definitely see your point for exposing a port from docker for manual testing. 

My concern is how users are expected to use this feature. As far as I remember just running `python -m http.server` wasn't enough (cf e.g. https://github.com/iodide-project/pyodide/issues/132)  because one also needs to set CORS and/or MIME WASM type (set [here](https://github.com/iodide-project/pyodide/blob/3c87663e1f5ea24b9108ee5d2c8fee53f84eb1d5/test/conftest.py#L337-L341) in `test/conftest.py`). Unless it's working for you without it? And if one does need to run `python tests/conftest.py` we need to make it optionally serve on port 8080 in this use case. Might have been nice to expose it as an entry-point `./bin/pyodide serve` (or something similar) as well.

**Comment by mdboom** [comment]
Thanks.  I should have looked at this in more detail.

I think @rth's suggestion of having an easy entry point to run the server is a good one.

That would probably involve:

- [ ] Add a new module in `pyodide_build` with a function to properly start up the server on the correct port for access outside of Docker
- [ ] Hook that function up as a command line subcommand [here](https://github.com/iodide-project/pyodide/blob/3c87663e1f5ea24b9108ee5d2c8fee53f84eb1d5/pyodide_build/__main__.py#L13)
- [ ] Documenting how to call it

@madhur-tandon : Are you interested in taking this on?

**Comment by madhur-tandon** [comment]
@mdboom Sure, I would like to work on this.
Just one request though, Can I start working on this after 4th December?
I have my end semester exams and projects to wrap up for the deadlines at my University.
Thanks!

**Comment by mdboom** [comment]
@madhur-tandon : No problem!

**Comment by madhur-tandon** [comment]
@rth @mdboom This is my first attempt at adding an entry point for running the server. Can you guys review this? I have added both ``--build_dir`` and ``--port`` as an option for the script. Let me know about this. Thanks!

File: pyodide_build/serve.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    return httpd
+
+
+def main(args):
+    build_dir = args.build_dir
+    port = args.port
+    httpd = server(port)
+    try:
+        os.chdir(build_dir)
+        print("serving from {0} at localhost:".format(build_dir) + str(port))
```
Comment (mdboom) [inline pyodide_build/serve.py]: I don't think these two lines need to be inside the try block.

**Comment by madhur-tandon** [comment]
Sure, Will fix that.
Lastly, I think that since we are adding an option to manually set the port which is to be used by the server inside the docker (using ``--port`` argument in ``serve.py``). Let's say the user passes in ``--port 6000``, then running the docker with ``-p 8000:8000`` does not make sense. It should be run with ``-p 8000:6000`` [8000 of local machine always, but 6000 of the docker environment] instead. Thus, I was thinking to add arguments to the shell script ``run_docker`` too, something like ``-p 8000:$1``. Then, the user could actually decide which port is to be used inside the docker environment passing the same values for the shell script ``run_docker`` as well as ``serve.py``. What about this @mdboom @rth ?

File: pyodide_build/serve.py (hunk 1 above)
Comment (madhur-tandon) [reply pyodide_build/serve.py]: Should I remove the ``except KeyboardInterrupt`` block in that case?

**Comment by rth** [comment]
> Thus, I was thinking to add arguments to the shell script run_docker too, something like -p 8000:$1. Then, the user could actually decide which port is to be used inside the docker environment passing the same values for the shell script run_docker as well as serve.py

I would just keep things simple, define some default port, and let user edit `run_docker` if they want another one. It's more a developer feature anyway.

**Comment by rth** [comment]
I agree this looks good (assuming some documentation is added). Ideally, I was hoping we could move some of the server functionality from `test/conftest.py` to `pyodide_build/serve.py`. Though it looks like doing so would not actually simplify things that much because most things are somewhat duplicate but slightly different, so this is probably sufficient..

**Comment by madhur-tandon** [comment]
@rth @mdboom Does the documentation I wrote suffice? Also, I have done some minor changes. I think this seems okay. What do you think?

File: README.md [hunk 2]
```diff
@@ -64,6 +64,22 @@ on your `PATH`.
 
 `pytest test/`
 
+# Manual Testing
+
+The port 8000 of the docker environment and the system are automatically binded when ``./run_docker`` is run.
+
+This can be used to test the ``pyodide`` builds running within the docker environment using external browser
+programs of the system.
```
Comment (mdboom) [inline README.md]: maybe, to be clearer: "using a browser on the host system"
